src/charm.py

```diff
@@ -561,29 +582,101 @@ def _on_mongod_pebble_ready(self, event) -> None:
 
         except (PathError, ProtocolError, MissingSecretError) as e:
             logger.error("Cannot initialize workload: %r", e)
-            event.defer()
-            return
+            return False
 
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("mongod", self._mongod_layer, combine=True)
-        container.add_layer("log_rotate", self._log_rotate_layer, combine=True)
+        return True
+
+    def __configure_layers(self, container: Container) -> bool:
+        """Configure the layers of the container."""
+        modified = False
+        current_layers = container.get_plan()
+        new_layers = {
+            Config.SERVICE_NAME: self._mongod_layer,
+            "log_rotate": self._log_rotate_layer,
+        }
         if self.is_role(Config.Role.CONFIG_SERVER):
-            container.add_layer("mongos", self._mongos_layer, combine=True)
+            new_layers["mongos"] = self._mongos_layer
 
+        # Add Pebble config layers missing or modified
+        for layer_name, definition in new_layers.items():
+            for service_name, service in definition.services.items():
+                if current_layers.services.get(service_name) != service:
+                    modified = True
+                    logger.debug(f"Adding layer {service_name}.")
+                    container.add_layer(layer_name, definition, combine=True)
+
+        # We'll always have a logrotate configuration at this point.
         container.exec(["chmod", "644", "/etc/logrotate.d/mongodb"])
 
-        # Restart changed services and start startup-enabled services.
-        container.replan()
+        return modified
+
+    def _configure_container(self, container: Container):
+        """Configure MongoDB pebble layer specification."""
+        if not container.can_connect():
+            logger.debug("mongod container is not ready yet.")
+            raise ContainerNotReadyError
+
+        # We need to check that the storages are attached before starting the services.
+        # pebble-ready is not guaranteed to run after storage-attached so this check allows
+        # to ensure that the storages are attached before the pebble-ready hook is run.
+        if any(not storage for storage in self.model.storages.values()):
+            logger.debug("Storages are not attached yet")
+            raise ContainerNotReadyError
+
+        if not self.__filesystem_handler(container):
+            raise ContainerNotReadyError
+
+        if self.__configure_layers(container):
+            container.replan()
```

Comment:
I think it is safe enough to always replan, like our conversation earlier

Reply:
I think it's not and as I encountered the issue with pbm returning 1 because of a premature restart, I'd rather be safe

Reply:
Do you know what causes the pre-mature restart, or what the error is? I think it would prob be better to prevent/delay the restart of pbm, than to compute whether the pebble layer has changed (something that pebble replan does for us)